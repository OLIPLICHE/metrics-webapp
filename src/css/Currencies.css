.btn {
  width: 100%;
  margin: 0.5rem 0;
  padding: 0.5rem;
  border: none;
  font-size: 1rem;
  font-weight: bold;
  color: inherit;
  background-color: var(--color-primary-dark);
  cursor: pointer;
}

.btn:hover {
  background-color: var(--color-primary-darker);
}

.header-tile {
  display: flex;
  align-items: center;
  height: var(--tile-size);
  background-color: var(--color-primary-light);
}

.header-tile > * {
  flex: 1;
  padding: 0.5rem;
}

.header-tile > :first-child {
  text-transform: uppercase;
  font-size: 4rem;
  color: var(--color-primary-darker);
}

.header-tile > :last-child {
  font-size: 2rem;
}

.stats-header {
  padding: 0.5rem;
  font-size: 0.8rem;
  font-weight: bold;
  text-transform: uppercase;
  background-color: var(--color-primary-dark);
}

.tiles-wrapper {
  display: flex;
  flex-wrap: wrap;
  width: 100%;
}

.tiles-wrapper a:nth-child(4n+1) > div,
.tiles-wrapper a:nth-child(4n) > div {
  background-color: var(--color-primary);
}

.item-tile {
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: var(--tile-size);
  width: var(--tile-size);
  padding: 0.5rem;
  background-color: var(--color-primary-dark);
}

.item-tile-illustration {
  font-size: 4rem;
  color: var(--color-primary-darker);
  user-select: none;
}

.item-arrow-icon {
  position: absolute;
  top: 0.5rem;
  right: 0.5rem;
  font-size: 2rem;
}

.item-tile-content {
  text-align: end;
}

.item-tile-content > h2 {
  text-transform: uppercase;
}

.item-tile-content > h3 {
  font-weight: normal;
}
