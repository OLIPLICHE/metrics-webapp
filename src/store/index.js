import {
  combineReducers,
  createStore,
  applyMiddleware,
} from 'redux';
import thunk from 'redux-thunk';

import rootReducer from '../reducers.js';

const reducers = combineReducers({
  root: rootReducer,
});

const store = createStore(reducers, applyMiddleware(thunk));

export default store;
