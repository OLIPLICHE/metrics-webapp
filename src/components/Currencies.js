import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import store from '../redux/configStore.js';
import { combineCurrencyRate, filterRates } from '../items/items.js';

import '../css/Currencies.css';

const Currencies = () => {
  const currencies = useSelector((state) => state.currencies);
  const usdExchange = useSelector((state) => state.usdExchange);

  const [rates, setRates] = useState([]);

  useEffect(() => {
    setRates(combineCurrencyRate(currencies, usdExchange));
  }, []);

  store.subscribe(() => {
    setRates(combineCurrencyRate(currencies, usdExchange));
  });

  const handleClick = (filter) => {
    setRates(filterRates(rates, filter));
  };

  const generateTiles = (rates) => {
    const tiles = rates.map((rate) => (
      <NavLink to={`/currency/${rate.code}`} key={rate.code}>
        <div className="item-tile">
          <span className="material-icons-outlined item-arrow-icon">
            <i className="fal fa-arrow-circle-right" />
          </span>
          <div className="item-tile-illustration">
            <h1>{rate.code}</h1>
          </div>
          <div className="item-tile-content">
            <h2>{rate.name}</h2>
            <h3>{rate.value}</h3>
          </div>
        </div>
      </NavLink>
    ));
    return tiles;
  };

  return (
    <main>
      <div className="header-tile">
        <div>
          <h1>1 USD</h1>
        </div>
        <div>
          <button
            className="btn"
            type="button"
            onClick={() => handleClick('alphabetical')}
          >
            alphabetical
          </button>
          <button
            className="btn"
            type="button"
            onClick={() => handleClick('most_valued')}
          >
            most valued
          </button>
          <button
            className="btn"
            type="button"
            onClick={() => handleClick('least_valued')}
          >
            least valued
          </button>
        </div>
      </div>
      <div className="stats-header">
        Stats by Currency
      </div>
      <div className="tiles-wrapper">
        {generateTiles(rates)}
      </div>
    </main>
  );
};

export default Currencies;
