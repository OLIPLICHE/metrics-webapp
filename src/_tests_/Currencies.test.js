import TestRenderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom/extend-expect.js';
import store from '../redux/configStore.js';
import Currencies from '../components/currencies.js';

jest.mock('../api/currency-api.js');

describe('Integration tests in Currencies', () => {
  it('matches the snapshot', () => {
    const tree = TestRenderer.create(
      <Provider store={store}>
        <Currencies />
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders filter buttons', () => {
    render(
      <Provider store={store}>
        <Currencies />
      </Provider>,
    );
    const filterBtn = screen.getByText(/alphabetical/i);
    expect(filterBtn).toBeInTheDocument();
  });

  it('Renders the reference currency', () => {
    render(
      <Provider store={store}>
        <Currencies />
      </Provider>,
    );
    const referenceCurrency = screen.getByText(/usd/i);
    expect(referenceCurrency).toBeInTheDocument();
  });
});
