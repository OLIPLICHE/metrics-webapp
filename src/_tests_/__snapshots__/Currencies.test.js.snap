// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests in Currencies matches the snapshot 1`] = `
<main>
  <div
    className="header-tile"
  >
    <div>
      <h1>
        1 USD
      </h1>
    </div>
    <div>
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        alphabetical
      </button>
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        most valued
      </button>
      <button
        className="btn"
        onClick={[Function]}
        type="button"
      >
        least valued
      </button>
    </div>
  </div>
  <div
    className="stats-header"
  >
    Stats by Currency
  </div>
  <div
    className="tiles-wrapper"
  />
</main>
`;
